### PR TITLE
ivy: add @ operator modifier

### PR DESCRIPTION
**Pull request by rsc**
The @ operator modifier is a bit like APL's each/map.
Applied to the left or right of an operator, it produces an operator
that applies to each element of its left or right argument and then
returns an array of those results.

For example consider a vector of x,y points and the operator
computing the Manhattan distance between two points:

	d = 3 2 rho iota 6

	op x dist y = +/ abs x-y

	0 0 dist d
	3 7 11

What if we want the Manhattan distance between all pairs of points?
This is a kind of outer product but only matching pairs of points;
the actual outer product matches all pairs of numbers.

The outer product does 'for all' over all dimensions of its
left and right arguments, while @ only does 'for all' over a
single dimension of the argument on the side where the @ appears.
So the pairwise Manhattan distances of the points is @dist@:

	d @dist@ d
	0 4 8
	4 0 4
	8 4 0

What if we have a binary operator and want to apply it with
a fixed left argument to each of a list of left or right arguments?
In this example, because Manhattan distance only uses +/, abs, and -,
and those automatically replicate a vector to match with a matrix,
it works to say:

	0 0 dist d
	3 7 1

	d dist 1 1
	1 5 9

but in general it would not. In the general case, dist@ and @dist work:

	0 0 dist@ d
	3 7 11

	d @dist 1 1
	1 5 9

What if we have an operator to compute pairwise distances
of a list of points and want to apply it to a list of point sets?
Then we can use a single @ on the right
(for a unary operator, it would never make sense to use @ on the left):

	op dists d = d @dist@ d

	dx = 4 3 2 rho iota 24

	dists@ dx
	0 4 8
	4 0 4
	8 4 0

	0 4 8
	4 0 4
	8 4 0

	0 4 8
	4 0 4
	8 4 0

	0 4 8
	4 0 4
	8 4 0

Outer product can be viewed as a special case of @.
Specifically, x o.f y is x @...@f@..@ where there are
(rho rho x) @s on the left and (rho rho y) @s on the right.

	(2 rho iota 3) o.^ 2 2 rho iota 4
	0 3
	2 5

	3 0
	1 6

	(2 rho iota 3) @^@@ 2 2 rho iota 4
	0 3
	2 5

	3 0
	1 6

### CODE REVIEW

**Comment by rsc** [comment]
Not for review. To support conversation in #69.

**Comment by robpike** [comment]
I'd need to see the documentation.

**Comment by tcolgate** [comment]
If we swtiched `@` to `each` and added docs, (in line with the discussion in #69), would this PR be acceptable?

**Comment by robpike** [comment]
The rank operator applies to "functions" as a first-class data type, which Ivy does not have. I'd like it to have them but it's a big change internally and I've never managed to make it happen. It may happen one day. I still think about it.

**Comment by rsc** [comment]
Proposed documentation is:

```
Operators and axis indicator

	Name                APL  Ivy  APL Example  Ivy Example  Meaning (of example)
	Reduce (last axis)  /    /    +/B          +/B          Sum across B
	Reduce (first axis) ⌿    /%   +⌿B                       Sum down B
	Scan (last axis)    \    \    +\B          +\B          Running sum across B
	Scan (first axis)   ⍀    \%   +⍀B                       Running sum down B
	Inner product       .    .    A+.×B        A +.* B      Matrix product of A and B
	Outer product       ∘.   o.   A∘.×B        A o.* B      Outer product of A and B
	                                                        (lower case o; may need preceding space)
	Each left                @op               A @op B      (A[1] op B), (A[2] op B), ... 
	                                                        as vector or matrix
	Each right          op¨  op@  A op¨ B      A op@ B      (A op B[1]), (A op B[2]), ... 
	                                                        as vector or matrix
```

**Comment by rsc** [comment]
> These seem fairly similar to K's [each-left and each-right](https://code.kx.com/q/ref/maps/#each-left-and-each-right), .... Were you aware of those when you were designing these?

I was not aware of the K operators; thanks for the link. Earlier I was talking to Rob and started with "Left each" / "Right each" but decided it was nicer to call them "Each left" and "Each right". I am happy to see that K chose exactly the same names. I agree that the operations seem identical.

As far as syntax if you think about `@` as a kind of operator modifier denoting each, I find `@op`, `op@`, and `@op@` easier to parse and remember which is which, compared to `op\:`, `op/:`, and `op/:\:`. I suppose the colon is meant to evoke an umlaut.

**Comment by rsc** [comment]
Pushed a new version with the docs we discussed, tests, and a [new commit message with better examples](https://github.com/robpike/ivy/pull/83/commits/b83ac1d432db533da72bef42c91a21fd2913bd84).

**Comment by rsc** [comment]
Playing with tryapl a bit more, it looks like I partly misunderstood what APL's each does. It looks like it is more special than just what this PR calls `f@`. Instead it seems to auto-adapt. For vector v and scalar s:

```
v f¨ s => v @f s
s f¨ v => s f@ v
v f¨ v => v @f@ v
```

But then if you have two different length vectors or an operation that wants a vector on one side, you can't use f¨ at all. The each-left and each-right distinction in in this PR and in K seem strictly more powerful.



```
TryAPL Version 3.6.1 (enter ]State for details)
Sat Dec 07 2024 08:08:00
Copyright (c) Dyalog Limited 1982-2024
      1 2 3 ,¨ 0⍴1
LENGTH ERROR
      1 2 3,¨0⍴1
           ∧
      1 2 3 ,¨ 4
┌───┬───┬───┐
│1 4│2 4│3 4│
└───┴───┴───┘
      1 ,¨ 2 3 4
┌───┬───┬───┐
│1 2│1 3│1 4│
└───┴───┴───┘
      1 2 3 ,¨ 4 5 6
┌───┬───┬───┐
│1 4│2 5│3 6│
└───┴───┴───┘
      1 2 3 ,¨ 4 5
LENGTH ERROR
      1 2 3,¨4 5
           ∧
```

**Comment by rsc** [comment]
Mix, split, and the general vector fixes made in 2023 (most notably a0c08b0577) postdate the design and implementation of this @ feature. In particular, I was working with an ivy like this:

```
(1 2) (3 4 5)
vector element must be scalar; have (1 2)
```

not like this:

```
(1 2) (3 4 5)
(1 2) (3 4 5)
```

The new, better vector support makes me think that probably @ should always return a vector, instead of forcing individual vector results into a matrix. Callers can always apply 'mix' if that's what they want. The current implementation does:

```
ivy@ '1 2 3' '4 5 6'
1 2 3
4 5 6

ivy@ '1 2 3' '4 5'
ivy@: conflicting result shapes (3) and (2)
```

Probably instead it should do:

```
ivy@ '1 2 3' '4 5 6'
(1 2 3) (4 5 6)

mix ivy@ '1 2 3' '4 5 6'
1 2 3
4 5 6

ivy@ '1 2 3' '4 5'
(1 2 3) (4 5)
```

I think that would be more in the spirit of the 2023 vector changes, and it is more useful and also a simpler implementation. I've run into that ivy@ problem a few times already.

I'll update the PR unless there are any objections.

File: doc.go [hunk 1]
```diff
@@ -202,7 +202,13 @@ Operators and axis indicator
 	Scan (first axis)   ⍀    \%   +⍀B                       Running sum down B
 	Inner product       .    .    A+.×B        A +.* B      Matrix product of A and B
 	Outer product       ∘.   o.   A∘.×B        A o.* B      Outer product of A and B
-	                                                       (lower case o; may need preceding space)
+	                                                        (lower case o;
+	                                                        may need preceding space)
+	Each left                @f                A @f B       (A[1] f B), (A[2] f B), ...
+	                                                        as vector or matrix;
+	                                                        may need preceding space
```
Comment (robpike) [inline doc.go:209]: why would it need a preceding space?

File: exec/operator.go [hunk 2]
```diff
@@ -4,7 +4,9 @@
 
 package exec
 
-import "robpike.io/ivy/value"
+import (
+	"robpike.io/ivy/value"
+)
```
Comment (robpike) [inline exec/operator.go:9]: restore layout

File: scan/scan.go [hunk 3]
```diff
@@ -329,22 +329,31 @@ func lexSpace(l *Scanner) stateFn {
...
 	}
+	for l.peek() == '@' {
+		l.next()
+	}
 	if !l.atTerminator() {
 		return l.errorf("bad character %#U", l.next())
 	}
 	// Some identifiers are operators.
 	word := l.input[l.start:l.pos]
+	w := strings.Trim(word, "@")
```
Comment (robpike) [inline scan/scan.go:346]: you know where it is, you could compute it as you go through the @s  at line 335 and 338.
i suppose it's not important as lexOperator doesn't have that luxury. your choice.

in any case both here and below "w" is a lousy name. maybe "word" and "opWord"

File: scan/scan.go [hunk 4]
```diff
@@ -672,6 +682,22 @@ func isAllDigits(s string, base int) bool {
 // isOperator reports whether r is an operator. It may advance the lexer one character
 // if it is a two-character operator.
 func (l *Scanner) isOperator(r rune) bool {
+	n := 0
+	for r == '@' {
+		r = l.next()
+		n++
+	}
+	if !l.isOperator1(r) {
```
Comment (robpike) [inline scan/scan.go:690]: isOperatorToken

File: scan/scan.go [hunk 5]
```diff
@@ -672,6 +682,22 @@ func isAllDigits(s string, base int) bool {
...
+		l.pos -= n
+		return false
+	}
+	for l.peek() == '@' {
+		l.next()
+	}
+	return true
+}
+
+func (l *Scanner) isOperator1(r rune) bool {
```
Comment (robpike) [inline scan/scan.go:700]: isOperatorToken

File: value/eval.go [hunk 6]
```diff
@@ -470,6 +470,143 @@ func ScanFirst(c Context, op string, v Value) Value {
...
+	}
+	return NewMatrix(mshape, result)
+}
+
+// Map computes the result of mapping op onto v.
+// The trailing @ has been removed.
+func Map(c Context, op string, v Value) Value {
+	var shape []int
+	var result Vector
+	var n int
```
Comment (robpike) [inline value/eval.go:561]: rather C-like. i don't use 'var' much. also i usually compute in []Value and convert to vector at the end. if that makes sense here,  please convert it all. 

shape := []int{}
n := 0
data := []Value{}

Comment (dmitshur) [reply value/eval.go:561]: Interestingly, that's different from https://go.dev/s/style#declaring-empty-slices which suggests the var style for empty slices.

File: doc.go (hunk 1 above)
Comment (rsc) [reply doc.go:209]: I thought it was an intentional feature of the lexer, but it was just a bug. Fixed.

File: exec/operator.go (hunk 2 above)
Comment (rsc) [reply exec/operator.go:9]: Fixed. I must have added a print and then when I deleted it and goimports removed the "fmt" import, it left the block. I gave up on trying to convince the goimports owners to change this behavior many years ago and just defer to the tool now. But fixed.

File: scan/scan.go (hunk 3 above)
Comment (rsc) [reply scan/scan.go:346]: Fixed; now there's just word, and it's trimmed.

File: scan/scan.go (hunk 4 above)
Comment (rsc) [reply scan/scan.go:690]: Done.

File: scan/scan.go (hunk 5 above)
Comment (rsc) [reply scan/scan.go:700]: Done.

File: value/eval.go (hunk 6 above)
Comment (rsc) [reply value/eval.go:561]: Now there's just one. I left it as Vector because it is now always a Vector result.

**Comment by rsc** [comment]
I forgot that GitHub doesn't do draft comments. I have made the changes in my local copy but am still working through changing the operator not to force the use of matrices as the result for things like ivy@ sys 'read' 'file'.

**Comment by rsc** [comment]
Sent an updated version. The rule for @ is now that it constructs only the dimensions that it took apart. 
For `x f@ y`, the result is a vector of x f y[i] results.
For `x @f y`, the result is a vector of x[i] f y results.
For `x f@@ y` the result is a matrix of x f y[i;j] results.
For `x @@f y`, the result is a matrix of x[i;j] f y results.
For `x @f@ y` the result is a matrix of x[i] f y[j] results.
And so on.

I updated all my uses of @ in this year's Advent of Code (including as-yet-untweeted solutions) and I think the result is cleaner.

And now things like this work instead of being rejected:

```
iota@ iota 5
(1) (1 2) (1 2 3) (1 2 3 4) (1 2 3 4 5)
```

**Comment by rsc** [comment]
Pushed a corrected commit message. Now `@f@` and `o.f` are an exact match, suggesting we are on a good track.

A few bugs remain to fix, but I'm stopping for tonight:

```
% ivy
op x shift y = x

d=1 2

d@shift 1
undefined global variable "d@"

1@shift 2
panic: impossible eachValue [recovered]
	panic: impossible eachValue

goroutine 1 [running]:
robpike.io/ivy/run.Run.func1()
	/Users/rsc/src/robpike.io/ivy/run/run.go:72 +0x294
panic({0x104601460?, 0x10462bdf0?})
	/Users/rsc/go/src/runtime/panic.go:787 +0x124
robpike.io/ivy/value.eachValue({0x10462d738?, 0x1045f00e8}, 0x10471a600?)
	/Users/rsc/src/robpike.io/ivy/value/eval.go:542 +0x1e8
```

**Comment by rsc** [comment]
Fixed those two bugs, both (surprisingly) rooted in the scanner.

**Comment by rsc** [comment]
I removed the use of Go 1.23 range over func, but I left the code prepared for it.

File: scan/scan.go [hunk 7]
```diff
@@ -346,11 +356,17 @@ func lexIdentifier(l *Scanner) stateFn {
 		return lexOperator
 	case l.defined(word):
 		return lexOperator
+	case word == "" || eachLeft > 0:
+		return l.errorf("invalid operator %v", l.input[l.start:l.pos])
```
Comment (robpike) [inline scan/scan.go:360]: %q

File: value/eval.go [hunk 8]
```diff
@@ -470,6 +470,178 @@ func ScanFirst(c Context, op string, v Value) Value {
...
+	case Vector:
+		return []int{len(v)}
+	case *Matrix:
+		return v.Shape()
+	}
+	return []int{}
+}
+
+// TODO: When we update to Go 1.23, Edit ,s/iterSeq/iter.Seq[Value]/g
+type iterSeq[T any] func(yield func(T) bool)
```
Comment (robpike) [inline value/eval.go:486]: i asked for this not to happen now. not only are you adding for sequencing, you're adding generics to this codebase.

**Review by robpike — CHANGES_REQUESTED** [review]

File: value/eval.go (hunk 8 above)
Comment (rsc) [reply value/eval.go:486]: It simplifies BinaryEach and Each considerably to abstract away the three choices about what's being iterated over from the code doing the iterating. My original implementation converted everything to Matrix and then computed a stride and passed subslices on each side, but that turned out to contain a subtle bug: `scalar foo@ v` turned the scalar left argument into a 1-element vector when invoking foo. This silently worked in my programs in 2021 but caused confusing errors in the new (better!) more-vector-preserving Ivy of 2024. To fix the accidental coercion, I need to handle the nine possible cases (scalar, vector, matrix) x (scalar, vector, matrix) with separate code. Cutting it into one function called twice with three cases is considerably nicer than attempting to spell them out.

I can keep the call sites that simple without generics (I only need `func(func(Value)bool)` not `iterSeq[Value]`) and with a slightly simpler callback function (I really only need `func(func(Value))`). Happy to do that if you want, but it seems strange to pretend not to know what the standard callback type will be.

Comment (robpike) [reply value/eval.go:486]: OK, but only if you promise to help me apply the approach universally. I do not like half updates.

Comment (rsc) [reply value/eval.go:486]: Will do.

File: scan/scan.go (hunk 7 above)
Comment (rsc) [reply scan/scan.go]: Done.

**Comment by rsc** [comment]
I think I've resolved all the comments now.

**Pull request merged**